src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+                }
+
+                return null;
+            }
+            else
+            {
+                // condition ? consequence : alternative
+                //
+                // becomes
+                //
+                // GotoIfFalse condition alt;
+                // capture = consequence
+                // goto afterif;
+                // alt:
+                // capture = alternative;
+                // afterif:
+                // result - cupture
+
+                SpillEvalStack();
+                int captureId = _availableCaptureId++;
+
+                BasicBlock whenFalse = null;
+                VisitConditionalBranch(operation.Condition, ref whenFalse, false);
+
+                AddStatement(new SimpleAssignmentExpression(new FlowCapture(captureId, operation.Syntax, 
+                                                                            isInitialization:true, operation.Type, 
+                                                                            default(Optional<object>)),
+                                                            operation.IsRef,
+                                                            Visit(operation.WhenTrue),
+                                                            semanticModel:null,
+                                                            operation.WhenTrue.Syntax,
+                                                            operation.Type,
+                                                            default(Optional<object>),
+                                                            isImplicit:true));
+
+                var afterIf = new BasicBlock(BasicBlockKind.Block);
+                LinkBlocks(CurrentBasicBlock, afterIf);
+                _currentBasicBlock = null;
+
+                AppendNewBlock(whenFalse);
+
+                AddStatement(new SimpleAssignmentExpression(new FlowCapture(captureId, operation.Syntax,
+                                                                            isInitialization: true, operation.Type,
+                                                                            default(Optional<object>)),
+                                                            operation.IsRef,
+                                                            Visit(operation.WhenFalse),
+                                                            semanticModel: null,
+                                                            operation.WhenFalse.Syntax,
+                                                            operation.Type,
+                                                            default(Optional<object>),
+                                                            isImplicit: true));
+
+                AppendNewBlock(afterIf);
+
+                return new FlowCapture(captureId, operation.Syntax, isInitialization:false, operation.Type, operation.ConstantValue);
+            }
+        }
+
+        private void SpillEvalStack()
```

Comment:
>SpillEvalStack [](start = 21, length = 14)

This is scanning the eval stack (possibly more than once) to create `FlowCapture`s and statement, and skipping those that are already done.
I'm wondering whether it would be simpler to create those as operations are pushed onto the eval stack, rather than later.
Would that mess up the order of `AddStatement` calls?

Reply:
> This is scanning the eval stack (possibly more than once) to create FlowCaptures and statement, and skipping those that are already done.
I'm wondering whether it would be simpler to create those as operations are pushed onto the eval stack, rather than later.
Would that mess up the order of AddStatement calls?

Spilling is done only when it is needed. If we were unconditionally creating "captures" all the time, we wouldn't need the eval stack.

---
In reply to: [162174434](https://github.com/dotnet/roslyn/pull/24263#discussion_r162174434) [](ancestors = 162174434)